[MLIR][TORCH] Add lowering for `aten._softmax` when `half_to_float=True`

Abhishek-Varma opened the pull request:
-- This commit adds decompose logic for `aten._softmax` when
   `half_to_float` is `True`.
-- I've added an MLIR lit test.
-- An e2e test case will be added once support for half to float conversion for
    `aten._softmax` is added.

Signed-off-by: Abhishek Varma <abhishek@nod-labs.com>

pashu123 reviewed (approved):
LGTM!

silvasean commented:
Please remove the .mlir test and add an end-to-end test. See here for our policy: https://github.com/llvm/torch-mlir/blob/main/docs/architecture.md#dos-and-donts-for-unit-vs-end-to-end-testing

pashu123 commented:
> Please remove the .mlir test and add an end-to-end test. See here for our policy: https://github.com/llvm/torch-mlir/blob/main/docs/architecture.md#dos-and-donts-for-unit-vs-end-to-end-testing

@silvasean The e2e test for this op can't be added for the torch's cpu version and requires cuda version (float16 version of softmax). @Abhishek-Varma can add a test though, but it will be ultimately marked as xfail.

silvasean commented:
Ah, got it. I would remove the .mlir test then and just mention that in the commit description.

Abhishek-Varma commented:
Hi @silvasean - I've addressed your review comment.
1. Removed the .mlir test.
2. Updated the commit description to reflect the reason behind not adding an e2e test.

You may re-review/approve the PR.

silvasean commented on lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:
```diff
@@ -354,16 +353,26 @@ class DecomposeAten_SoftmaxOp : public OpRewritePattern<Aten_SoftmaxOp> {
+    BaseTensorType resultTensorType = op.getType().cast<BaseTensorType>();
+    // `torch.ops.aten._softmax`'s softmax with half to float conversion is not
+    // supported on CPU.
+    // But we go ahead with the decomposing and add a unit MLIR lit test.
```
this comment seems out of date.
